Add instructions for uploading to GitHub from RStudio (#10)
esciencecenter-digital-skills/reproducible-research-through-reusable-code: episodes/github.md

---
title: "Pre-workshop assignment: Uploading a coding project to GitHub"
teaching: 5
exercises: 45
---

::: questions
-   How do I share my changes with others on the web?
:::

::: objectives
-   Create a repository on GitHub
-   Push to or pull from a remote repository.
:::

> ## Creating a GitHub repository

Reviewer: Remove, only keep creating an account.

> You are going to add your existing project to GitHub

Reviewer: Expand further, make it clear what the goal is of this assignment. Explain a little bit why we put our code on GitHub. Add instructions for if your project is already on GitHub.

Next, you need to initialize git (version control) for your Project:

In Visual Studio Code, open the folder of the project that you want to add to GitHub. 
Then click on 'Source Control' and 'Initialize Repository':

![](fig/git-vscode.png){alt="Initializing version control in Visual Studio Code"}

Now, you can stage files by clicking on the '+' next to 'Changes' or next to the individual files:

![](fig/git-vscode-stage.png){alt="Staging files in Visual Studio Code"}

An 'A' will denote that the files are now added, and they are in the category 'Staged Changes':

![](fig/git-vscode-added-staged.png){alt="Staged files in Visual Studio Code"}

Then you click on 'Commit'. Then you are asked to add a so-called commit message. When you accept and save
that message, you can publish the remote on GitHub by clicking the blue 'Publish Branch' button that should now
be available.  

When you do so you are asked to sign in to GitHub and authorize Visual Studio Code via the browser. After that, you
will return to Visual Studio Code and you should click 'Publish to public GitHub Repository'.

> Refresh the home page of your repository on GitHub to verify that your code is there.

Reviewer: ```suggestion
The home page of your repository on GitHub should now automatically open int your browser. Verify that your code is there.
```

Next, you need to initialize git (version control) for your Project:

In Rstudio, open the [RStudio Project](https://support.posit.co/hc/en-us/articles/200526207-Using-RStudio-Projects) that you want to add to GitHub. Go to 'Tools \> Version control \> Project setup...'

Select 'Git' as the Version Control System

![](fig/git-rstudio.png){alt="Initializing version control in RStudio Project"}

Then confirm that you want to initialize a new git repository for the project.

'Confirm New Git Repository' and then confirm that you want to restart RStudio.

After restarting RStudio, your project is reopened and a 'Git' Version Control tab is added.

If you click on it, you see the following:

![](fig/git-tab-rstudio.png){alt="The Git tab in an RStudio Project"}

The next step is to add all your files to be “monitored”.

*Note that RStudio automatically adds files you want to ignore in a .gitignore file. You can add certain files (e.g. sensitive data that should not be uploaded) yourself if needed.*

> In the Git tab, you can tick the files that you want to upload. Then you click on 'Commit pending changes'.

Reviewer: ```suggestion
In the Git tab, you can tick the files that you want to upload. Then you click on 'Commit (pending changes)'.
```

Reviewer: For me it only shows 'commit'.

A new window will open. Here you can add a so-called commit message.

> Use something descriptive, in this case for example "Initial commit" and click on 'Commit'.

Reviewer: ```suggestion
Use something descriptive, in this case for example "Initial commit" and click on 'Commit'. A new pop-up window might open with a summary of your commit. You can close it.
```

> Now, you want to link to your project on GitHub. The home page of the repository on GitHub includes the URL string we need to identify:

Reviewer: ```suggestion
So far, we put some changes together in a commit, but we have not 'pushed' these changes to GitHub. Let us link to your project on GitHub. The home page of the repository on GitHub includes the URL string we need to identify:
```

![](fig/github-find-repo-string.png){alt="Where to Find Repository URL on GitHub"}

Make sure to copy the HTTPS link and not the SSH link.

> -   **Linux and Mac**

Reviewer: ```suggestion
### Linux and Mac
```

Author: I think this will make it an extra tab (next to RStudio and VSCode), that is why I did not use ## here

On a linux-based machine (including Mac OS), you should be able to go back to RStudio, and open a new Terminal window:

Go to Tools in the menu bar, select Terminal, and then New Terminal

Then paste in the three commands from the GitHub page, into the terminal window and press enter:

``` bash
git remote add origin https://github.com/yourusername/my-awesome-project.git
git branch -M main
git push -u origin main
```

Refresh the home page of your repository on GitHub to verify that your code is there.

> -   **Windows**

Reviewer: ```suggestion
### Windows
```

On a windows machine, set RStudio to open Git Bash terminals instead of Command Prompt or Windows Powershell.

Go to Tools in the menu bar and select 'Terminal Options'.

Select Git Bash from the drop down menu, select Apply, then OK.

You can then open a new terminal window by going to 'Tools \> Terminal \> New Terminal'.

In the new terminal window, paste in the three commands from the GitHub page and press enter:

``` bash
git remote add origin https://github.com/yourusername/my-awesome-project.git
git branch -M main
git push -u origin main

> ```

Reviewer: I think this is really terrible that you still have to do this and there is now option to configure this through the RStudio interface. But I think this is the only way to set the origin main branch as upstream branch. If you've done this once you can use the push and pull buttons in the Rstudio git interface

> Refresh the home page of your repository on GitHub to verify that your code is there.

Reviewer: ```suggestion
Refresh the home page of your repository on GitHub to verify that your code is there.

Note that from now on you can also use the 'push' and 'pull' buttons in the `Git` tab to upload and download changes to your project on GitHub.
```

git commit -m "Initial Commit"
```

Now, we want to link to your project on GitHub. The home page of the repository on GitHub includes the URL string we need to identify it:

![](fig/github-find-repo-string.png){alt="Where to Find Repository URL on GitHub"}

Make sure to copy the HTTPS link and not the SSH link.

Then use below command to connect the local repository to the repository in GitHub

```         
git remote add origin <project url>
```

Test to see that it worked by doing

``` bash
git remote -v
```

You should see what your repo is linked to.

Then you can push your changes to GitHub

``` bash
git push origin main
```

Refresh the home page of your repository on GitHub to verify that your code is there.
:::
::::

::: callout
## Wait, what did we just do?

> These are the basics for uploading a project to GitHub. We realize we are skipping a lot of details on how git works and how to use it. Our excuse is we want reproducible code on GitHub within a day.

Reviewer: Great disclaimer!

Author: I think you wrote it in the previous command line instructions haha

:::

::: challenge
## Optional exercise: My code is already on GitHub

If your code is already on GitHub you can try to help others pushing their code to GitHub, or explore the following topics:

-   [Familiarize yourself with the basics of `git`](https://swcarpentry.github.io/git-novice/){target="_blank"}
-   [Learn more about `.gitignore` files](https://swcarpentry.github.io/git-novice/06-ignore.html){target="_blank"}
-   If you already know the basics of `git`, familiarize yourself with best practices in using git with [this lesson](https://carpentries-incubator.github.io/python-intermediate-development/14-collaboration-using-git/index.html){target="_blank"}. This lesson assumes you have some project with changes to it, you can make some changes in the project you are working on today to mimic the lesson.
:::

::: keypoints
-   Use GitHub in the browser to create a remote repository
-   Use `git init` to initialize a local repository
-   Use `git add .` to add all your files to be "monitored" by git
-   USe `git commit` to commit your changes
-   Use `git push` to upload your local project to GitHub

> :::

Reviewer: These should be updated!
